packages/api/src/external/carequality/document/query-documents.ts

```diff
@@ -0,0 +1,213 @@
+import { out } from "@metriport/core/util/log";
+import dayjs from "dayjs";
+import pTimeout from "p-timeout";
+import { DocumentReference as FHIRDocumentReference } from "@medplum/fhirtypes";
+import { S3Utils, createS3FileName } from "@metriport/core/external/aws/s3";
+import { errorToString } from "@metriport/core/util/error";
+import { DocumentQueryResult } from "../domain/document-query-result";
+import { DocumentReference } from "../domain/ihe-result";
+import { Patient } from "../../../domain/medical/patient";
+import { getDocumentQueryResult } from "../command/ihe-result/get-document-query-result";
+import { Config } from "../../../shared/config";
+import { makeFhirApi } from "../../fhir/api/api-factory";
+import { getAllPages } from "../../fhir/shared/paginated";
+import { isConvertible } from "../../fhir-converter/converter";
+import { appendDocQueryProgress } from "../../../command/medical/patient/append-doc-query-progress";
+import { createCQDocumentQueryRequest } from "./document-query-request";
+import { getOrganizationOrFail } from "../../../command/medical/organization/get-organization";
+import { makeIheGatewayAPI } from "../api";
+import { capture } from "../../../shared/notifications";
+import { MedicalDataSource } from "../../../external";
+import { getCQPatientData } from "../command/cq-patient-data/get-cq-data";
+import { MAPIWebhookStatus } from "../../../command/medical/document/document-webhook";
+import { processPatientDocumentRequest } from "../../../command/medical/document/document-webhook";
+// import { downloadDocs } from "./download-documents";
+
+const s3Utils = new S3Utils(Config.getAWSRegion());
+const s3BucketName = Config.getMedicalDocumentsBucketName();
+const iheGateway = makeIheGatewayAPI();
+
+export const DOCUMENT_QUERY_TIMEOUT = dayjs.duration({ minutes: 2 });
+
+export async function getDocumentsFromCQ({
+  requestId,
+  patient,
+}: {
+  requestId: string;
+  patient: Patient;
+}) {
+  if (!iheGateway) return;
+
+  const { log } = out(`CQ query docs - requestId ${requestId}, M patient ${"123"}`);
+
+  try {
+    const organization = await getOrganizationOrFail({ cxId: patient.cxId });
+    const cqPatientData = await getCQPatientData({ id: patient.id, cxId: patient.cxId });
+
+    const docQueryRequest = createCQDocumentQueryRequest({
+      requestId,
+      cxId: patient.cxId,
+      organization,
+      cqLinks: cqPatientData?.data.links ?? [],
+    });
+
+    await iheGateway.startDocumentsQuery(docQueryRequest);
+
+    // TODO: CHANGE THIS TO THE STRATEGY RAMIL USED: RACE CONTROL
+    const results = await pTimeout(
+      getDocumentQueryResult({ requestId }),
+      DOCUMENT_QUERY_TIMEOUT.asMilliseconds(),
+      "handleDocQueryResponse function timed out!"
+    );
```

Comment:
`getDocumentQueryResult()` just queries the DB, why do we need something like pTimeout or Promise.race?

And, if I understand correctly what we're trying to achieve here, it seems we need to break this operation: we shouldn't have the API waiting for minutes for an operation, even its not consuming CPU - we have restarts!

This is a job for delayed/scheduled tasks: https://github.com/metriport/metriport-internal/issues/1315

Could look into alternatives, I just think this should be broken in two.

Reply:
Created lambda for this